README.md
# (MODX)EvolutionCMS.plugins.ManagerManager.mm_ddMultipleFields

Widget for plugin ManagerManager that allows you to add any number of fields values (TV) in one document (values are written in field as a JSON object). For example: a few images.

Capabilities:
* Save several images, text fields, selects, etc to a document field.
* Create several columns with different (or identical) data types. For example: images with titles (`$params->columns`).
* Number of rows may be fixed, dynamic or may lay in the special range (`$params->minRowsNumber`, `$params->maxRowsNumber`).
* Rows sorting by drag and drop.
* Generating of an unique ID for each row.
* List of predefined values for a column (`$params->columns[i]['type']` == `'select'`).


## Requires

* PHP >= 5.4
* [(MODX)EvolutionCMS.plugins.ManagerManager](https://code.divandesign.biz/modx/managermanager) >= 0.7


## Documentation


### Installation

To install you must unzip the archive to `/assets/plungins/managermanager/widgets/`.


You may also read this documentation:
* [(MODX)EvolutionCMS.plugins.ManagerManager](https://code.divandesign.biz/modx/managermanager).
* [(MODX)EvolutionCMS.modules.ddMMEditor](https://code.divandesign.biz/modx/ddmmeditor).
* [(MODX)EvolutionCMS.snippets.ddGetMultipleField](https://code.divandesign.biz/modx/ddgetmultiplefield).


Type of TV must be `textarea`.


### Parameters description

* `$params`
	* Desctription: Parameters, the pass-by-name style is used.
	* Valid values:
		* `stdClass`
		* `arrayAssociative`
	* **Required**
	
* `$params->fields`
	* Desctription: Names of TV for which the widget is applying.
	* Valid values: `stringCommaSeparated`
	* **Required**
	
* `$params->columns`
	* Desctription: Columns.
	* Valid values: `array`
	* Default value: `[ ['type' => 'text'] ]`
	
* `$params->columns[i]`
	* Desctription: Column.
	* Valid values: `arrayAssociative`
	* **Required**
	
* `$params->columns[i]['type']`
	* Desctription: Column type.
	* Valid values:
		* `'text'` — `<input type="text">` column
		* `'textarea'` — `<textarea>` column
		* `'richtext'` — column with rich text editor
		* `'image'` — image column
		* `'file'` — file column
		* `'date'` — date column
		* `'select'` — `<select>` column (see `$params->columns[i]['data']`)
	* **Required**
	
* `$params->columns[i]['title']`
	* Desctription: Column title.
	* Valid values: `string`
	* Default value: `''`
	
* `$params->columns[i]['width']`
	* Desctription: Column width, px.
	* Valid values: `integer`
	* Default value: `180`
	
* `$params->columns[i]['data']`
	* Desctription: Valid values (for the `'select'` column type)
	* Valid values: `stringJsonArray`
	* Default value: —
	
* `$params->columns[i]['data'][i]`
	* Desctription: Item.
	* Valid values: `arrayAssociative`
	* **Required**
	
* `$params->columns[i]['data'][i]['value']`
	* Desctription: Item value.
	* Valid values: `string`
	* **Required**
	
* `$params->columns[i]['data'][i]['title']`
	* Desctription: Item title.
	* Valid values: `string`
	* Default value: == `$params->columns[i]['data'][i]['value']`
	
* `$params->minRowsNumber`
	* Desctription: Minimum number of rows.
	* Valid values: `integer`
	* Default value: `0`
	
* `$params->maxRowsNumber`
	* Desctription: Maximum number of rows.
	* Valid values:
		* `integer`
		* `0` — unlimited
	* Default value: `0`
	
* `$params->previewWidth`
	* Desctription: Maximum value of image preview width (for the `'image'` columns).
	* Valid values: `integer`
	* Default value: `300`
	
* `$params->previewHeight`
	* Desctription: Maximum value of image preview height (for the `'image'` columns).
	* Valid values: `integer`
	* Default value: `100`
	
* `$params->roles`
	* Desctription: The CMS user roles that the widget is applied to.
	* Valid values:
		* `stringCommaSeparated`
		* `''` — when this parameter is empty then widget is applied to the all roles
	* Default value: `''`
	
* `$params->roles[i]`
	* Desctription: CMS user role.
	* Valid values: `integer`
	* **Required**
	
* `$params->templates`
	* Desctription: Document templates IDs for which the widget is applying to.
	* Valid values:
		* `stringCommaSeparated`
		* `''` — empty value means the widget is applying to all templates
	* Default value: `''`
	
* `$params->templates[i]`
	* Desctription: Templates ID.
	* Valid values: `integer`
	* **Required**


### CMS events

* `OnDocFormPrerender`
* `OnDocFormRender`


### Output format

The widget save value to a TV as JSON object with the following structure:

```json
{
	"1590412453247": {
		"0": "First row, first column value",
		"1": "First row, second column value"
	},
	"1590412497589": {
		"0": "Second row, first column value",
		"1": "Second row, Second column value"
	}
}
```

Where:
* `1590412453247`, `1590412497589` — the unique auto generated row IDs (JS `(new Date).getTime()` is used while creating rows).
* `0`, `1` — column numbers.

Rows objects with empty column values will not be saved. If all columns and all rows are empty, an empty string (`''`) will be saved instead of an empty JSON ojbect (`'{}'`).

It is strongly recommend to use [(MODX)EvolutionCMS.snippets.ddGetMultipleField](https://code.divandesign.biz/modx/ddgetmultiplefield) >= 3.5 for rendering TVs on site.


### Examples


#### Make the TV `someImages` available for adding several number of images

Create the TV `someImages`, set it's type equal to `textarea`.

```php
mm_ddMultipleFields([
	'fields' => 'someImages',
	'columns' => [
		//Only one column
		[
			'type' => 'image'
		]
	]
]);
```


#### Create 2 columns: images with titles

```php
mm_ddMultipleFields([
	'fields' => 'someImage',
	'columns' => [
		[
			'type' => 'image',
			'title' => 'Photo'
		],
		[
			'type' => 'text',
			'title' => 'Title'
		]
	]
]);
```


#### Table of employees contacts

Create the TV `employees`, set it's type equal to `textarea`.

```php
mm_ddMultipleFields([
	'fields' => 'employees',
	'columns' => [
		[
			'type' => 'text',
			'title' => 'Name',
			'width' => 250
		],
		[
			'type' => 'text',
			'title' => 'Phone'
			'width' => 100
		],
		[
			'type' => 'text',
			'title' => 'Job title'
			'width' => 200
		]
	],
	//Minimum 2 employee is required
	'minRowsNumber' => 2,
	//And maximum 10
	'maxRowsNumber' => 10
]);
```


#### `<select>` column type

```php
mm_ddMultipleFields([
	'fields' => 'employees',
	'columns' => [
		[
			'type' => 'text',
			'title' => 'Name',
			'width' => 250
		],
		[
			'type' => 'text',
			'title' => 'Phone'
			'width' => 100
		],
		[
			'type' => 'select',
			'title' => 'Job title'
			'data' => '[
				{
					"value": "Founder"
				},
				{
					"value": "SEO"
				},
				{
					"value": "Designer"
				},
				{
					"value": "Product manager"
				},
				{
					"value": "Developer"
				}
			]'
		]
	]
]);
```


## Links

* [Home page](https://code.divandesign.biz/modx/mm_ddmultiplefields)
* [Telegram chat](https://t.me/dd_code)


<link rel="stylesheet" type="text/css" href="https://DivanDesign.ru/assets/files/ddMarkdown.css" />

CHANGELOG.md
# (MODX)EvolutionCMS.plugins.ManagerManager.mm_ddMultipleFields changelog


## Version 4.8.2 (2020-05-26)
* \* Richtext column: **All** `<`, `>` and `&` are not replaced to HTML entities (`&lt;`, `&gt;` and `&amp;` respectively).


## Version 4.8.1 (2020-05-25)
* \* Improved uniqueness row ID while it is used for backward compatibility.
* \* `jQuery.ddMM.mm_ddMultipleFields.init`: Fixed deprecated ID column using.


## Version 4.8 (2020-05-25)
* \* Richtext column: `<`, `>` and `&` are not replaced to HTML entities (`&lt;`, `&gt;` and `&amp;` respectively).
* \* TV value stores as a JSON object, no more separated strings (with backward compatibility).
* \+ The widget always generates an unique ID for each row and save it as a row key in result object (see README).
* \* Parameters → `$params->columns[i]['type']`: The `'id'` value is deprecated (with backward compatibility).
* \* It is strongly recommend to use (MODX)EvolutionCMS.snippets.ddGetMultipleField >= 3.5 for rendering TVs on site.
* \+ Empty rows (rows objects with empty column values) will not be saved.
* \+ README.
* \+ README_ru.
* \+ CHANGELOG.
* \+ CHANGELOG_ru.
* \+ Composer.json.


## Version 4.7.4 (2019-06-23)
* \* New path of (MODX)EvolutionCMS.libraries.ddTools for compatibility with the new ManagerManager version.


## Version 4.7.3 (2018-11-22)
* \* Style: Improved input height.
* \* The `$modx->getConfig` method is used instead of the `$modx->config` property.
* \* Richtext pupup window (_Many thanks to [@MrSwed](https://github.com/MrSwed)._):
	* \+ Textarea autofocus.
	* \* No scrolls in the window.
	* \* Correct window and element sizes.


## Version 4.7.2 (2018-03-29)
* \* `jQuery.ddMM.mm_ddMultipleFields`:
	* \* Datepicker now gets the value of day and months from MODX config.
	* \* `makeSelect`: Set field width using the parameter.


## Version 4.7.1 (2017-05-03)
* \* Richtext editor is used from `$modx->config['which_editor']`. _Many thanks to [@MrSwed](https://github.com/MrSwed)._
* \* Rich text field in popup window is now working correctly even TinyMCE is not installed.


## Version 4.7 (2016-11-19)
* \* Attention! PHP >= 5.4 is required.
* \* Attention! (MODX)EvolutionCMS.plugins.ManagerManager >= 0.7 is required.
* \* PHP short array syntax is used because it's more convenient.
* \* Refactoring, the widget now using named parameters (with backward compatibility).
* \* Refactoring: `jQuery.ddMM.fields` is used istead of manually DOM finding.
* \* `jQuery.ddMM.mm_ddMultipleFields`:
	* \* Refactoring, the plugin now using named parameters.
	* \* Refactoring `jQuery.ddMM.mm_ddMultipleFields.instances`:
		* \* Items is pushed by `jQuery.ddMM.mm_ddMultipleFields.init`.
		* \* `currentField` was renamed as `$currentField`.
		* \+ `$parent`, `$originalField` and `$table` ware added.
* \* The `$params->columns` parameter now must be an array. Items contain type, title, width and data. `$params->columnsTitles`, `$params->columnsWidth` and `$params->columnsData` are deprecated. Of course, with backward compatibility.
* \+ Add button is inserted in every row (closes #7).
* \* Small style changes.


## Version 4.6 (2014-10-24)
* \* Attention! (MODX)EvolutionCMS.plugins.ManagerManager >= 0.6.3 is required.
* \* Be advised! You should use the `image` and / or `file` column types instead of `field` (backward compatibility maintained).
* \* The `$dir` variable was renamed as `$richtextIncludeDirectory` because of some namespace troubles.
* \* The `textarea` input type is now required for TVs with multiple fields regardless of what their field types are.
* \* `$.ddMM.mm_ddMultipleFields` has been updated to 1.2:
	* \* You should use the `image` and / or `file` column types instead of `field`.
	* \- The parameters `makeFieldFunction` and `browseFuntion` were removed as unnecessary.
	* \- The method `maskQuoutes` has been removed. Therefore, quotes are not now converted to html-entities.
	* \* Values of inputs are assigned via `$.fn.val`.
	* \* Added initializations of the following functions `OpenServerBrowser`, `BrowseServer`, `BrowseFileServer`, `SetUrlChange`, `SetUrl` if required (it's a copy from the MODX kernel).
	* \* Modification of `SetUrl` is performed only for old versions of MODX (if the function `SetUrlChange` is absent).


## Version 4.5.1 (2014-05-15)
* \* `jQuery.fn.mm_ddMultipleFields` has been updated to 1.0.1:
	* \* Removing events of the (MODX)EvolutionCMS.plugins.ManagerManager.mm_widget_showimagetvs plugin has been added.
* \* `richtext/index.php`:
	* \* The redundant slash has been removed from the include statements.
	* \* The `manager/includes/protect.inc.php` file is icluded before `manager/includes/config.inc.php`.
	* \* The relative path and `MGR_DIR` are used in include statements instead of `MODX_MANAGER_PATH` (needed if the site is not located in `$_SERVER['DOCUMENT_ROOT']`).


## Version 4.5 (2013-12-10)
* \* Attention! (MODX)EvolutionCMS.plugins.ManagerManager >= 0.6 is required.
* \+ The new `richtext` type of `$coloumns` has been added.
* \+ Support of the `textarea` and `email` TV types has been added.
* \* The three calls of `tplUseTvs` were replaced by the single one with arguments.
* \* JS and CSS files are currently included via `includeJsCss` to avoid occuring of duplicates.
* \* The JS code became a stand-alone file and was partially revised. It is convenient because of PHP code cleanness and it shortens the amount of code of a document edit frame.
* \* The including of required JS and CSS files is currently being realized while occuring the event `OnDocFormPrerender`. The files are included as HTML.
* \* Special chars are currently escaped during parsing of `columnsData` regardless of eval use.
* \* The title and width of the column `id` are set as `''` and `0` respectively due to redundancy of the data because the column is hidden.

ddmultiplefields/ddmultiplefields.php
<?php
/**
 * mm_ddMultipleFields
 * @version 4.8.1 (2020-05-25)
 * 
 * @link https://code.divandesign.biz/modx/mm_ddmultiplefields
 * 
 * @copyright 2012–2020 DD Group {@link http://DivanDesign.biz }
 */

function mm_ddMultipleFields($params){
	//For backward compatibility
	if (
		!is_array($params) &&
		!is_object($params)
	){
		//Convert ordered list of params to named
		$params = \ddTools::orderedParamsToNamed([
			'paramsList' => func_get_args(),
			'compliance' => [
				'fields',
				'roles',
				'templates',
				'columns',
				'columnsTitles',
				'columnsWidth',
				'rowDelimiter',
				'colDelimiter',
				'previewWidth',
				'previewHeight',
				'minRowsNumber',
				'maxRowsNumber',
				'columnsData'
			]
		]);
	}
	
	//Defaults
	$params = (object) array_merge(
		[
			'fields' => '',
			'columns' => [
				[
					'type' => 'text'
				]
			],
			'minRowsNumber' => 0,
			'maxRowsNumber' => 0,
			'previewWidth' => 300,
			'previewHeight' => 100,
			'roles' => '',
			'templates' => '',
			//Deprecated
			'rowDelimiter' => '||',
			'colDelimiter' => '::',
			'columnsTitles' => '',
			'columnsWidth' => '180',
			'columnsData' => ''
		],
		(array) $params
	);
	
	if (!useThisRule(
		$params->roles,
		$params->templates
	)){
		return;
	}
	
	global $modx;
	
	$e = &$modx->Event;
	
	$output = '';
	
	$site = $modx->getConfig('site_url');
	$widgetDir =
		$site .
		'assets/plugins/managermanager/widgets/ddmultiplefields/'
	;
	
	if ($e->name == 'OnDocFormPrerender'){
		$output .= includeJsCss(
			(
				$site .
				'assets/plugins/managermanager/js/jquery-ui-1.10.3.min.js'
			),
			'html',
			'jquery-ui',
			'1.10.3'
		);
		$output .= includeJsCss(
			(
				$widgetDir .
				'ddmultiplefields.css'
			),
			'html'
		);
		$output .= includeJsCss(
			(
				$widgetDir .
				'jQuery.ddMM.mm_ddMultipleFields.js'
			),
			'html',
			'jQuery.ddMM.mm_ddMultipleFields',
			'2.5.6'
		);
		
		$e->output($output);
	}else if ($e->name == 'OnDocFormRender'){
		global $mm_current_page;
		
		$params->fields = tplUseTvs(
			$mm_current_page['template'],
			$params->fields,
			'image,file,text,email,textarea',
			'type,name'
		);
		
		if ($params->fields == false){
			return;
		}
		
		//Колонки, заданные как «field», теперь их нужно будет заменить на «image» и «file» соответственно
		$columns_fieldKeyIndex = [];
		
		//Old columns format backward compatibility
		if (!is_array($params->columns)){
			$columnsTemp = makeArray($params->columns);
			$params->columnsTitles = makeArray($params->columnsTitles);
			$params->columnsWidth = makeArray($params->columnsWidth);
			
			//Prepare data
			if ($params->columnsData){
				$columnsDataTemp = explode(
					'||',
					$params->columnsData
				);
				$params->columnsData = [];
				
				foreach (
					$columnsDataTemp as
					$dataItem
				){
					//For backward compatibility '[["Value 1", "Title 1"], ["Value 2"]]' → '[{"value" => "Value 1", "title" => "Title 1"}, {"value" => "Value 2"}]'
					if ($dataItem != ''){
						$dataItemTemp = json_decode(
							$dataItem,
							true
						);
						$dataItem = [];
						
						//Build list
						foreach (
							$dataItemTemp as
							$dataItem_item_index =>
							$dataItem_item_value
						){
							$dataItem[$dataItem_item_index] = [];
							$dataItem[$dataItem_item_index]['value'] = $dataItem_item_value[0];
							
							//Title
							if (isset($dataItem_item_value[1])){
								$dataItem[$dataItem_item_index]['title'] = $dataItem_item_value[1];
							}
						}
						
						$dataItem = json_encode($dataItem);
					}
					
					$params->columnsData[] = $dataItem;
				}
			}else{
				$params->columnsData = [];
			}
			
			$params->columns = [];
			
			foreach (
				$columnsTemp as
				$index =>
				$value
			){
				//“field” value compatibility
				if ($value == 'field'){
					$columns_fieldKeyIndex[] = $index;
				}
				
				$value = ['type' => $value];
				
				if (isset($params->columnsTitles[$index])){
					$value['title'] = $params->columnsTitles[$index];
				}
				if (isset($params->columnsWidth[$index])){
					$value['width'] = $params->columnsWidth[$index];
				}
				if (isset($params->columnsData[$index])){
					$value['data'] = $params->columnsData[$index];
				}
				
				$params->columns[] = $value;
			}
		}
		
		//Default value for columns
		if (empty($params->columns)){
			$params->columns = [
				['type' => 'text']
			];
		}
		
		//Стиль превью изображения
		$previewStyle =
			'max-width:' .
			$params->previewWidth .
			'px; max-height:' .
			$params->previewHeight .
			'px; margin: 4px 0; cursor: pointer;'
		;
		
		$output .=
			'//---------- mm_ddMultipleFields :: Begin -----' .
			PHP_EOL
		;
		
		foreach (
			$params->fields as
			$field
		){
			//For backward compatibility
			if (
				$field['type'] == 'image' ||
				$field['type'] == 'file'
			){
				//Проходимся по всем колонкам «field» и заменяем на соответствующий тип
				foreach(
					$columns_fieldKeyIndex as
					$index
				){
					$params->columns[$index]['type'] = $field['type'];
				}
			}
			
			$output .=
'
$j.ddMM.fields.' . $field['name'] . '.$elem.mm_ddMultipleFields({
	columns: ' . json_encode(
		$params->columns,
		JSON_UNESCAPED_UNICODE
	) . ',
	previewStyle: "' . $previewStyle . '",
	minRowsNumber: "' . $params->minRowsNumber . '",
	maxRowsNumber: "' . $params->maxRowsNumber . '",
	
	rowDelimiter: "' . $params->rowDelimiter . '",
	colDelimiter: "' . $params->colDelimiter . '"
});
';
		}
		
		//Поругаемся
		if (!empty($columns_fieldKeyIndex)){
			$modx->logEvent(
				1,
				2,
				'<p>You are currently using the deprecated column type “field”. Please, replace it with “image” or “file” respectively.</p><p>The plugin has been called in the document with template id ' . $mm_current_page['template'] . '.</p>',
				'ManagerManager: mm_ddMultipleFields'
			);
		}
		
		$output .=
			'//---------- mm_ddMultipleFields :: End -----' .
			PHP_EOL
		;
		
		$e->output($output);
	}
}
?>